### ISSUE TITLE
Type inference of Intersection Types goes wrong.

### ISSUE THREAD

**Issue by biqas** [Microsoft/TypeScript#8894]
**TypeScript Version:** 

1.8.2

**Code**

```ts
// A self-contained demonstration of the problem follows...
someFunction<T extends ClassA & ClassB>(x : T & ClassB) : T {}

let x : ClassA & ClassB & ClassC;
let result = someFunction(x); // => Is not inferring correctly.
```

**Expected behaviour:**
result should be: "ClassA & ClassB & ClassC".

**Actual behaviour:** 
result is: "ClassA & ClassB".

**Comment by biqas** [comment]
Of course it does not make sense do declare it in this way, but if it is really the case then there should be a warning or error?!

**Comment by ahejlsberg** [comment]
Inference actually works as expected in your example. Below I have removed the type parameter constraint:

```typescript
declare function someFunction<T>(x: T & ClassB) : T;
let x : ClassA & ClassB & ClassC;
let result = someFunction(x);
```

This infers `ClassA & ClassC` for `T` as expected. However, when you add the constraint `ClassA & ClassB` you end up with an inference for `T` that doesn't meet the constraint (specifically, `ClassA & ClassC` is not assignable to `ClassA & ClassB`). When an inferred type isn't assignable to its constraint, we go with the constraint instead, so the resulting inference becomes `ClassA & ClassB`. Since the argument `x` *is* assignable to that, there is no error. As a rule we don't generate errors from the type inference process--it is simply supposed to come up with a best possible match, and then errors (if any) are generated by the assignability checks between arguments and parameters.

I'm not exactly sure what you're trying to accomplish in your example, but I'd suggest either removing the constraint or changing the parameter to `x: T` instead of `x: T & ClassB`.

**Comment by biqas** [comment]
Hi Anders,

actually the function signature was produced by accident not intentionally so no special purpose behind it :), I was just wondering, why the result is not the same as the input?! I can understand that in the function it is constrained by the rules you described, but if it flows out, it should be the same or not?

My understanding of inference here was, a type T flows in has some constraints and becomes internally in the function scope T' (T-prime) you work internally with T' (T-prime) if type flows out inference flow switch back to T?!

**Comment by biqas** [comment]
A T goes in a T comes out, why should T change?! I'm not convinced why it should be this way?! :-)

**Comment by ahejlsberg** [comment]
Here's the reasoning: You're passing an `A & B & C` to something that takes a `T & B`. From that we infer that `T` must be `A & C`. Then we check if that inference meets the constraint `A & B` you've specified. It doesn't. Since a type argument is always required to meet its constraint (because you said so), we conclude that we can't infer a type for you and revert to the constraint.

Feel free to suggest a new set of rules, but I think what we're doing makes sense.

**Comment by biqas** [comment]
Hi Anders,

I totally believe that what you are doing makes sense! And now I also understand this rule better.

By the way I only choose to work intensiv with Typescript because of you! :)
I like the way you are designing and reasoning about languages and I have used quite a long time C# and still using it.

And this was what I had in my mind and thought it could be deduced by the compiler.

```typescript
declare function someFunction<T extends ClassA & ClassB, E>(
    a: T & ClassB, b : E) : T & E;

let x : ClassA & ClassB & ClassC;
let r = someFunction(x, x); // r : ClassA & ClassB & ClassC
```
My assumtion was:

Substitute direction:
That through reverse analysis in this case you could detect (x, x) is equivalent to (x), and input type and inferred type are equal therefor it follows that you could substitute E and b : E and T & E => T.

Expansion/Probing direction
In case constraints are not met and the given type (in this case: ClassA & ClassB & ClassC) is a intersection type of a certain cardinality, expand the function signature and try if substitution rules can apply.

As said this is just an assumtion and not proven, and also given the fact that this is not really the common scenario I do not know if it is worth. It was just a moment where I thought wait a second why is this not possible nothing more.
